### ISSUE TITLE
Crashes with filter-cross-facet transform pipeline with signals

### ISSUE THREAD

**Issue by Jerrythafast** [vega/vega#442]
I discovered multiple ways to crash Vega at [Line 30 of datalib/collector.js](https://github.com/vega/datalib/blob/47611b786822146b081311cfdfefd772c3ce8282/src/aggregate/collector.js#L30) with an "invalid Array length" RangeError.

MWE 1 crashes if you try `ved.view.signal("limit", 3).update()`:
```json
{
  "signals": [{"name": "limit", "init": 4}],
  "data": [
    {
      "name": "a",
      "values": [1, 2, 3],
      "transform": [
        {"type": "filter", "test": "datum.data < limit"},
        {"type": "cross"},
        {"type": "formula", "field": "groupie", "expr": "datum.b.data"},
        {"type": "facet", "groupby": ["groupie"]}
      ]
    }
  ]
}
```

MWE 2 crashes if you try `ved.view.signal("limit", 2).update()`:
```json
{
  "signals": [{"name": "limit", "init": 4}],
  "data": [
    {
      "name": "a",
      "values": [1, 2, 3],
      "transform": [
        {"type": "filter", "test": "datum.data < limit"},
        {"type": "cross"},
        {"type": "facet", "groupby": ["a"]}
      ]
    }
  ]
}
```

With #441 to make the `cross` behave correctly, MWE 2 crashes as well if you do
```js
ved.view.signal("limit", 3).update();
ved.view.signal("limit", 4).update(); //crash
```

Furthermore, if you load MWE 2 and then do `ved.view.data("a").values()` you'll find that it has created only one facet with `key: "[object Object]"`. I would have expected it to create three facets at first, because the `key`s are actually three different objects that only happen to have identical string representations. I'm almost tempted to suggest using `JSON.stringify()` and/or some hashing function if you require the `key` to be a string.

Kind regards,
Jerry

**Comment by arvind** [comment]
Our old friend, the Cross transform, is to blame. It is inserting duplicate tuples into `mod` and `rem`. Look for a fix soon. For your latter issue, regarding grouping by on an object, I would recommend opening an issue on the [datalib repo](https://github.com/vega/datalib/issues/new) so that we can track it distinctly from the bug faced here. Thanks!

**Comment by jheer** [comment]
Hi @Jerrythafast, regarding object-valued aggregation keys, these are not supported in datalib and I do not have plans to support them at this time. Among other things, I would be concerned about the possible performance implications of complex key generation. One option may be to list the properties you want to key on explicitly (e.g., `[a.foo, a.bar, a.baz]`), generate key strings yourself (e.g., via a formula transform) and group on those, or find other ways to "flatten" your tuples first. I suspect a number of use cases might be resolved through alternative data organizations. That said, you should of course feel free to submit issues to datalib, ideally with motivating examples that could not be easily handled through other means!

**Comment by Jerrythafast** [comment]
Hey @arvind, I'm not sure whether this is completely fixed. For both MWEs in this issue, Vega crashes if you do
```js
ved.view.signal("limit", 3).update();
ved.view.signal("limit", 4).update(); //crash
```
The crash happens when increasing the value of `limit` back to `4`.

**Comment by arvind** [comment]
Good catch. Fixed with a815cef.

**Comment by arvind** [comment]
Alright, besides additional bugs with Cross (a815cef, 2dfd6e1), the kicker was that the Aggregate relied on bookkeeping that only occurred in the render node of the dataflow graph. With only datasources specified, the render node was never called on subsequent signal updates. I've moved this logic to be called once a pulse propagation is complete (vega/vega-dataflow@47483db, v1.3.2).
